update: CMP tutorial, updates with the new default project structure (#637)
JetBrains/kotlin-multiplatform-dev-docs: topics/compose-onboard/compose-multiplatform-explore-composables.md

The `App()` function is a regular Kotlin function annotated with `@Composable`.
Such functions are referred to as _composable functions_ or simply _composables_.
They are the building blocks of a UI based on Jetpack Compose or Compose Multiplatform.

This `App()` function is used as the base of the UI architecture for the app and has the following structure:

* The `MaterialTheme()` sets the look of the application. The default settings can be customized. For example, you can
  choose colors, shapes, and typography.
* The `Column()` composable controls the layout of the application. Here, it displays a `Button` above
  the `AnimatedVisibility()` composable.
* The `Button()` contains the `Text` composable, which renders some text on top of the button.
* The `AnimatedVisibility()` call is set up to show and hide the `Image` using an animation when the button is pressed.
* The `painterResource()` loads a vector image stored as an XML file.

The `horizontalAlignment` parameter of the `Column()` function centers the column's content.
For this to come into effect, the column should take up the full width of its container.

> This is achieved using the `modifier` parameter.

Reviewer: ```suggestion
You can achieve this by using the `modifier` parameter.
```

Modifiers are a key component of Jetpack Compose and Compose Multiplatform.
They provide the primary mechanism to adjust the appearance or behavior of composables in the UI.
Modifiers are created using methods of the `Modifier` type. When you chain these
methods, each call can change the `Modifier` returned from the previous call, making the order significant.

> See the [Compose Multiplatform introduction to modifiers](https://kotlinlang.org/docs/multiplatform/compose-layout-modifiers.html#built-in-modifiers)

Reviewer: Can you swap this around? I think it's clearer to explain the why before the how.

and the extensive [Jetpack Compose modifier documentation](https://developer.android.com/jetpack/compose/modifiers) for more details.

## Managing the state

There is a persistent quality to the loaded image: it should consistently be either visible or hidden across recompositions

> unless the button is clicked.

Reviewer: ```suggestion
There is a persistent quality to the loaded image: it should consistently remain either visible or hidden across recompositions
unless user clicks the button.
```
Is the persistent quality "needed"? I feel like something is missing here.

The `showContent` property in the `App()` composable is built using the `mutableStateOf()` function,
which means it's a state object that can be observed:

```kotlin
var showContent by remember { mutableStateOf(false) }
```

The state object is wrapped in a `remember()` call, meaning that it's built once and then
retained by the framework. This way, the `showContent` property has a value that is a state object containing a boolean.
The framework caches this state object, allowing composables to observe it.

When the value of the state changes, any composables that observe it are re-invoked. This allows any of the widgets they
produce to be redrawn. This is called a _recomposition_.

The only place where the state is changed is in the `onClick` parameter of the `Button()` call.
The event handler flips the value of the `showContent` property.
As a result, the image gets shown or hidden along with a `Greeting().greet()` call
because the parent `AnimatedVisibility()` composable observes `showContent`.

## Launching UI on different platforms

The `App()` function is executed differently on each platform. On Android, it's managed by an activity; on iOS, by a view 

> controller; on the desktop, by a window; and on the web, by a container. Let's examine each of them.

Reviewer: ```suggestion
The `App()` function is executed differently on each platform:

* On Android, it's managed by an activity.
* On iOS, by a view controller.
* On the desktop, by a window.
* On the web, by a container.

Let's examine each of them.
```
What about making this a list?

```

This is an [Android activity](https://developer.android.com/guide/components/activities/intro-activities)
called `MainActivity` that invokes the `App()` composable declared in common code.

### On iOS

For iOS, open the `MainViewController.kt` file within `shared/src/iosMain/kotlin`:

```kotlin
fun MainViewController() = ComposeUIViewController { App() }
```

This is a [view controller](https://developer.apple.com/documentation/uikit/view_controllers) that performs the same
role as an activity on Android. Notice that both the iOS and Android types simply invoke the `App()` composable from common code.

### On desktop

For desktop, look for the `main.kt` file in `desktopApp/src/main/kotlin`:

```kotlin
fun main() = application {
    Window(
        onCloseRequest = ::exitApplication,
        title = "ComposeDemo"
    ) {
        App()
    }
}
```

* Here, the `application()` function launches a new desktop application.
  This function takes a lambda, which initializes the UI.

> * Typically, within `application()` you create a `Window` and specify its properties

Reviewer: ```suggestion
* Typically, within the `application()` function, you create a `Window` and specify its properties
```
